File: DailyLeetcoding/494-target-sum/target-sum.java
// class Solution {
//     public int findTargetSumWays(int[] nums, int target) {
//         if (nums.length == 1) {
//             if (nums[0] == target || -nums[0] == target) {
//                 return 1;
//             } else {
//                 return 0;
//             }
//         }
//         return findExps(nums, target, 0, 0);
//     }
    
//     private static int findExps(int[] nums, int target, int index, int sum) {
//         // base condition
//         if (index == nums.length) {
//             if (sum == target) {
//                 return 1;
//             }
//             return 0;
//         }
        
//         int count = 0;
//         // function call for + operator
//         count += findExps(nums, target, index + 1, sum + nums[index]);
        
//         // function call for - operator
//         count += findExps(nums, target, index + 1, sum - nums[index]);
        
//         return count;
//     }
// }


class Solution {
public int findTargetSumWays(int[] nums, int target) {

    int ans=allsets(nums, 0 , 0, target, 0);
    return ans;
}

public static int allsets(int[] nums,int idx, int sum, int target, int count){
    
    if(idx==nums.length){
        if(sum==target){
            count++;
        }
        return count;
    }
    
    int ansp=allsets(nums ,idx +1 , sum + nums[idx] ,target,count);
    
    int ansm=allsets(nums ,idx +1 , sum - nums[idx] ,target,count);

    return ansp +ansm;
}
}

File: DailyLeetcoding/494-target-sum/target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        memo = {} #memoizing the result to reduce recursive calls
        def rec (index, exp_result):
            # base case
            if index == len(nums):

                if exp_result == target:
                    return 1

                return 0

            if (index , exp_result) in memo:
                return memo[(index, exp_result)]

            result = rec(index + 1, exp_result + nums[index]) + rec(index + 1, exp_result - nums[index])
            memo[(index, exp_result)] = result

            return result

        return rec(0, 0)